hack/jenkins/upload-to-gcs.sh

```diff
@@ -40,19 +54,78 @@ readonly artifacts_path="${WORKSPACE}/_artifacts"
+  if [[ -e "version" ]]; then
+    version=$(cat "version")
+  elif [[ -e "hack/lib/version.sh" ]]; then
+    version=$(
+      export KUBE_ROOT="."
+      source "hack/lib/version.sh"
+      kube::version::get_version_vars
+      echo "${KUBE_GIT_VERSION-}"
+    )
   fi
-  # Mark this build as the latest completed.
-  echo "${BUILD_NUMBER}" | \
-    gsutil -q -h "Content-Type:text/plain" -h "Cache-Control:private, max-age=0, no-transform" \
-      cp -a "${gcs_acl}" - "${gcs_job_path}/latest-build.txt" || continue
-  break  # all uploads succeeded if we hit this point
-done
 
-readonly results_url=${gcs_build_path//"gs:/"/"https://storage.cloud.google.com"}
-echo -e "\n\n\n*** View logs and artifacts at ${results_url} ***\n\n"
+  if [[ -n "${version}" ]]; then
+    echo "Found Kubernetes version: ${version}"
+  else
+    echo "Could not find Kubernetes version"
+  fi
+
+  local -r json_file="${gcs_build_path}/started.json"
+  for upload_attempt in $(seq 3); do
+    echo "Uploading version to: ${json_file} (attempt ${upload_attempt})"
+    gsutil -q -h "Content-Type:application/json" cp -a "${gcs_acl}" <(
+      echo "{"
+      echo "    \"version\": \"${version}\","
+      echo "    \"timestamp\": ${timestamp}"
+      echo "}"
+    ) "${json_file}" || continue
+    break
+  done
+}
+
+function upload_artifacts_and_build_result() {
+  local -r build_result=$1
+  echo -n 'Run finished at '; date -d "@${timestamp}"
+
+  for upload_attempt in $(seq 3); do
+    echo "Uploading to ${gcs_build_path} (attempt ${upload_attempt})"
+    echo "Uploading build result: ${build_result}"
+    gsutil -q -h "Content-Type:application/json" cp -a "${gcs_acl}" <(
+      echo "{"
+      echo "    \"result\": \"${build_result}\","
+      echo "    \"timestamp\": ${timestamp}"
+      echo "}"
+    ) "${gcs_build_path}/finished.json" || continue
+    if [[ -d "${artifacts_path}" && -n $(ls -A "${artifacts_path}") ]]; then
+      echo "Uploading artifacts"
+      gsutil -m -q -o "GSUtil:use_magicfile=True" cp -a "${gcs_acl}" -r -c \
+        -z log,txt,xml "${artifacts_path}" "${gcs_build_path}/artifacts" || continue
+    fi
+    # Mark this build as the latest completed.
+    echo "Marking build ${BUILD_NUMBER} as the latest completed build"
+    echo "${BUILD_NUMBER}" | \
+      gsutil -q -h "Content-Type:text/plain" -h "Cache-Control:private, max-age=0, no-transform" \
+        cp -a "${gcs_acl}" - "${gcs_job_path}/latest-build.txt" || continue
+    break  # all uploads succeeded if we hit this point
+  done
+
+  echo -e "\n\n\n*** View logs and artifacts at ${results_url} ***\n\n"
+}
+
+if [[ -n "${JENKINS_BUILD_STARTED:-}" ]]; then
+  upload_version
+elif [[ -n "${JENKINS_BUILD_FINISHED:-}" ]]; then
+  upload_artifacts_and_build_result ${JENKINS_BUILD_FINISHED}
+else
+  echo "Called without JENKINS_BUILD_STARTED or JENKINS_BUILD_FINISHED set."
+  echo "Assuming a legacy invocation."
+  upload_artifacts_and_build_result "[UNSET]"
```

Comment:
also there may be a race between this merging + the Jenkins job configs updating.